src/validator.py

```diff
@@ -121,12 +123,12 @@ def validate_func_def(self, node, v_map, f_map, ctx):
             # TODO: support function overloading
             raise FunctionDefDuplication(
                 "Function {} already defined in this context".format(name))
+        f_map[name][ctx] = (_type, [v.type for v in node.args])
```

Comment:
Fix: Move function registration earlier to enable function recursion.

TODO: Does C allow calling a function defined after current place? If so we need multiple pass for validation.